src/platforms/hosted/bmp_libusb.c

```diff
@@ -47,291 +106,373 @@ void libusb_exit_function(bmp_info_s *info)
+							//
+							if (!probe_skip) {
+								useSerial = false;
+								probe_skip = strdup(product);
+							}
+						} else {
+							if (probe_skip) {
+								if (useSerial) {
+									if (strstr(serial, probe_skip))
+										add_probe = false; // Skip this interface
+								} else if (strstr(product, probe_skip))
+									add_probe = false;
+							}
+						}
+					}
+					if (add_probe)
+						probe_list = probe_info_add_by_id(
+							probe_list, BMP_TYPE_FTDI, NULL, vid, pid, manufacturer, product, serial, strdup("---"));
+					else {
+						free(serial);
+						free(product);
+						free(manufacturer);
+					}
+				}
+			}
+			if (probe_skip)
+				free(probe_skip);
+			free((void *)dev_info);
+		} else
+			DEBUG_WARN("process_ftdi_probe: memory allocation failed\n");
 	}
+	return probe_list;
+}
+#endif
 
-	libusb_device_handle *handle;
-	res = libusb_open(dev, &handle);
-	if (res != LIBUSB_SUCCESS) {
-		DEBUG_INFO("libusb_open() failed: %s\n", libusb_strerror(res));
-		libusb_free_config_descriptor(conf);
-		return type;
+void orbtrace_read_version(libusb_device *device, libusb_device_handle *handle, char *version, size_t buffer_size)
+{
+	libusb_config_descriptor_s *config;
+	if (libusb_get_active_config_descriptor(device, &config) == 0) {
+		for (size_t iface = 0; iface < config->bNumInterfaces; ++iface) {
+			const libusb_interface_s *interface = &config->interface[iface];
+			for (int descriptorIndex = 0; descriptorIndex < interface->num_altsetting; ++descriptorIndex) {
+				const libusb_interface_descriptor_s *descriptor = &interface->altsetting[descriptorIndex];
+				uint8_t string_index = descriptor->iInterface;
+				if (string_index == 0)
+					continue;
+				if (libusb_get_string_descriptor_ascii(handle, string_index, (unsigned char *)version, buffer_size) < 0)
+					continue; /* We failed but that's a soft error at this point. */
+
+				if (strstr(version, "Version") != NULL) {
+					break;
+				}
+			}
+		}
 	}
+}
 
-	for (uint8_t i = 0; i < conf->bNumInterfaces; ++i) {
-		const libusb_interface_descriptor_s *interface = &conf->interface[i].altsetting[0];
+bool process_cmsis_interface_probe(
+	libusb_device_descriptor_s *device_descriptor, libusb_device *device, probe_info_s **probe_list, bmp_info_s *info)
+{
+	(void)info;
+	char *serial;
+	char *manufacturer;
+	char *product;
+	char *version;
 
-		if (!interface->iInterface)
-			continue;
+	libusb_device_handle *handle;
+	bool cmsis_dap = false;
 
-		res = libusb_get_string_descriptor_ascii(
-			handle, interface->iInterface, (uint8_t *)interface_string, sizeof(interface_string));
-		if (res < 0) {
-			DEBUG_ERROR("libusb_get_string_descriptor_ascii() failed: %s\n", libusb_strerror(res));
-			continue;
-		}
+	libusb_config_descriptor_s *config;
 
-		if (!strstr(interface_string, "CMSIS"))
-			continue;
-		type = BMP_TYPE_CMSIS_DAP;
+	if (libusb_get_active_config_descriptor(device, &config) == 0 && libusb_open(device, &handle) == 0) {
```

Comment:
This procedure now leaks the config descriptor. Consider calling `libusb_free_config_descriptor(config)` at the end of routine before returning (and possibly in error handling early-returns if added later).
>SUMMARY: AddressSanitizer: 2555 byte(s) leaked in 39 allocation(s).

Same for `orbtrace_read_version`. Otherwise I see zero leaks from ASAN.

Reply:
This should now be taken care of.